core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -214,7 +210,8 @@ public void run() {
 
                     int plannedCapacitySnapshot = 0;
 
-                    for (Iterator<PlannedNode> itr = pendingLaunches.iterator(); itr.hasNext(); ) {
+                    List<PlannedNode> snapPendingLaunches = new ArrayList<PlannedNode>(pendingLaunches.get());
```

Comment:
I am actually considering switching to a `PlannedNode[]` rather than a `List<PlannedNode>` but I view that as being out of scope of the fix for the deadlock here.